dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/search/index/IndexServiceBean.java

```diff
@@ -300,6 +291,32 @@ public void asyncIndexDatasetList(List<Dataset> datasets, boolean doNormalSolrDo
         int numPublishedVersions = 0;
         List<DatasetVersion> versions = dataset.getVersions();
         List<String> solrIdsOfFilesToDelete = new ArrayList<>();
+        try {
+            /**
+             * Preemptively delete *all* Solr documents for files associated
+             * with the dataset based on a Solr query.
+             *
+             * We must query Solr for this information because the file has
+             * been deleted from the database ( perhaps when Solr was down,
+             * as reported in https://github.com/IQSS/dataverse/issues/2086
+             * ) so the database doesn't even know about the file. It's an
+             * orphan.
+             *
+             * @todo This Solr query should make the iteration above based
+             * on the database unnecessary because it the Solr query should
+             * find all files for the dataset. We can probably remove the
+             * iteration above after an "index all" has been performed.
+             * Without an "index all" we won't be able to find files based
+             * on parentId because that field wasn't searchable in 4.0.
+             *
+             * @todo We should also delete the corresponding Solr
+             * "permission" documents for the files.
+             */
+            List<String> allFilesForDataset = findFilesOfParentDataset(dataset.getId());
```

Comment:
Don't know If it's important in case of optimization or not. But I have strong feeling that `List<String> allFilesForDataset` will contain all of identifiers that is obtained below in:
```
            for (FileMetadata fileMetadata : fileMetadatas) {
                String solrIdOfPublishedFile = solrDocIdentifierFile + fileMetadata.getDataFile().getId();
```
Also `findFilesOfParentDataset` returns full solr documents. We can ask solr to retrieve only specific fields (we are only interested in ids).
There is also an option of making direct remove by query in solr instead of `getIds` - `removeIds` pattern

Reply:
Hmm. Maybe I wasn't clear enough. I think that:
`List<String> allFilesForDataset = findFilesOfParentDataset(dataset.getId());`
can return more ids (or the same amount) than
what is obtained via database by iterating over all FileMetadata in all DatasetVersion's
`String solrIdOfPublishedFile = solrDocIdentifierFile + fileMetadata.getDataFile().getId();`

For example we performed deletion of draft file - everything went fine 
So in result we should leave detection of ids to delete based on solr and remove detection by iteration over entities.

It is even mentioned in comment:
```
            * @todo This Solr query should make the iteration above based
             * on the database unnecessary because it the Solr query should
             * find all files for the dataset. We can probably remove the
             * iteration above after an "index all" has been performed.
             * Without an "index all" we won't be able to find files based
             * on parentId because that field wasn't searchable in 4.0.
```